### ISSUE TITLE
AnonymousType

### ISSUE THREAD

**Issue by justeff** [wojtpl2/ExtendedXmlSerializer#192]
It seems like that XmlTyp(AnonymousType = true) is not supported. 

```
        public class ClassA
        {
            public string ElementA { get; set; } = "TextA";
            public ClassB B { get; set; } = new ClassB();
        }
        
        [XmlType(AnonymousType = true)]
        public class ClassB
        {
            public string ElementB { get; set; } = "TextB";
        }
```

I want to prevent <B><ElementB>TextB</ElementB></B> and just get <ElementB>TextB</ElementB> inside ClassA. Is there another way to achieve this? I have to say, I don't like the standard Xml Serializer solution, because if I need a class C which should have Class B fully inside and not just his members, then I have to create a duplicate class for B but without the "AnonymousType" tag, that why I would favour a member tag like XmlElement to be more flexible.

**Comment by Mike-EEE** [comment]
Ah... well, unfortunately, we do not offer anything out of the box.  

Now, if there is only one property you need this for per class, then you could create an extension that could make a content property ala Xaml's [`ContentPropertyAttribute`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.markup.contentpropertyattribute).

This is of course something that I have been meaning to do because it makes a good extension story.  Unfortunately it's not in there, but we'll strive to make it for v3 (I really hate punting work, sorry!).

Of course, I am of the opinion that having property names is a nice thing to have in your serialized object graph. 😇

**Comment by justeff** [comment]
Well, you do not need to apologize :), I thought about making my own Xml Serializer after I created my own XSD code generator, but I guess it would be better to help you with your framework. I only have 2 rare cases where I would need this option. In the UML model, I have an aggregation and in the XSD file it's just a group reference and I thought it would be nice to hide the XSD definition from the user and just give him the UML representation (give him the group reference as the underlying class, but resolve it's members like if they belong to the root class (_ClassA_) ;)

**Comment by justeff** [comment]
Another question, are XML types supported like:

```
[XmlText(DataType = "unsignedInt")]
public uint Value { get; set; }
```

**Comment by Mike-EEE** [comment]
Ah cool... sure if you manage to build an extension we will gladly accept and review any PRs you have. :)  Do note that v2 (current version) is pretty much in release mode.  We do have a v3 planned but it's pretty much a rewrite. 

As for XML types... we do have our own system and derive it from CLR types, or the types defined in the member, if that makes sense.

**Comment by justeff** [comment]
Do you have an example for the XML types?

Because  _[XmlText(DataType = "dateTime")] public DateTime Value { get; set; } = DateTime.Now;_ and         _[XmlText(DataType = "date")] public DateTime Value { get; set; } = DateTime.Now;_ is not working for me. I would expect different output streams.

**Comment by Mike-EEE** [comment]
Ah, so what I was saying is that we have our own type system in this regard.  We always emit by default the value of a `DateTime` instance based on its registered converter, which is the result of `XmlConvert.ToDateTime`.  You can see this [here](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/master/src/ExtendedXmlSerializer/ContentModel/Conversion/DateTimeConverter.cs).

So, by default, all `DateTime`s and `DateTimeOffset`s look the same, if that makes sense.

If you want, this would be a good place to use a custom serializer on a member to yield the result you are looking to achieve. :)  Actually, it would be an extension that would work in conjunction with custom member serializers.

I should point out that while we do not support _every_ legacy/classic feature, the scope and intent of ExtendedXmlSerializer is to serialize/deserialize object graphs, rather than interop with other external data sources (which was classic XmlSerializer's primary concern, IMO).  Formatting text such as this is usually rooted in having to account for an external schema of some sort.  In this case, I would recommend using XSLT to get the data prepared correctly.

**Comment by justeff** [comment]
What's about [XmlText]? Seems like I have to stick with the standard XmlSerializer? :(

**Comment by Mike-EEE** [comment]
I guess I am not too familiar with `XmlText`.  Are you looking to format your `DateTime` in a specific way?  This can be done with extensions but there is nothing built yet for it, I'm afraid.

**Comment by Mike-EEE** [comment]
Ah alright... so it seems that this is indeed a mechanism to pull in content from an element and assign it to a property on a class.  Much like `ContentPropertyAttribute` as denoted earlier.  I am assuming the `dateTime` formats from the `DataType` property are used to properly read it in.

So what this is IMO is a ContentProperty, which could be read in via metadata (classic `XmlTextAttribute`) or via the `ConfigurationContainer`.

Maybe something like `new ConfigurationContainer().Type<Subject>().Content(x => x.Property)` and/or `new ConfigurationContainer().Type<Subject>().Member(x => x.Property).AsContent()`.

Is there anything specific/magical about the `XmlText` that is missing here?  I think everything else is covered by EXS's type system.

**Comment by justeff** [comment]
Yeah, a content property and no it's not DateTime, most of the time it's just a string oder List<string>. I have over 300 generated classes and about 70 of them use this kind of property. So using is no soultion for me and XmlText or XmlTextAttribute is not working, The property is treated like a regular element property ... hmm. YAX can handle this, but it has problems with collections of base classes (you don't have it). XmlSerializer can handle this all, but it's slow and you have set a lot of additional meta data informations when using inheritance and this is quite odd. Thanks for your help.

**Comment by Mike-EEE** [comment]
OK @justeff I ACCEPT YOUR CHALLENGE!!! 😆  Well, more like I feel your pain.  I myself have been spending the past week going through and _trying_ to get all my identities aligned between GOOG and MSFT... and there is not a single chat solution that I am happy with between the two.  This is especially perplexing as GOOG has like 5 different chat applications and each one of them does something good, but none of them do it exactly right.

So I did some research on this and I cannot find an example of XmlText used as a string list.  Do you have a quick example you can share?  I am mostly interested in the xml output.

**Comment by justeff** [comment]
Just realized, list string makes no sense for me. The XML is standardized, if I get it from an external source it will be just one big string haha. I need only a content property then :)

**Comment by Mike-EEE** [comment]
OK COOL... let's see what I can do. 👍

**Comment by Mike-EEE** [comment]
Alright... I checked this out and I think I have a reasonable path forward here.  I've got some tasks that I have to wrap up for today, but I am going to aim to have this available for you by EOD tomorrow if that works for you.  Assuming there aren't any super-gotchas that impede the path, that is. 😇

**Comment by justeff** [comment]
Thank you Mike, you don't need to hurry up, I'll use XmlSerializer to get my stuff done and maybe I can switch to this framework later on and benefit from peformance gains :) Have a good day.

**Comment by Mike-EEE** [comment]
OK cool... maybe I will push it to the weekend then. 😇  I'm getting ready to switch out my hyper-v server to a Windows 10 install and that is going to take a little time.  In either case, I will certainly keep this thread updated.

**Comment by Mike-EEE** [comment]
Holy moly! Have I been in a rabbit hole. I am just now getting my environment back online and being able to attend to this. I hope to have this resolved today or tomorrow. We'll see how I do. 😛

**Comment by Mike-EEE** [comment]
Whew... brutal.  Alright.  this sums up how I feel about taking this (self-imposed) challenge:

![](https://i.imgur.com/Q81RPrp.gif)

😛 

I got a rough draft in there for you to check out.  I'm not exactly happy with it, however.  It's a poor attempt at mimicking the classic serializer and it is fragile at best.  Unit tests pass but I doubt it will hold up to an actual real world scenario.

The good news is that it is all tucked away in an extension that has all of the components within it.  That's sort of the nice thing with this extension model: you can add something to the system that is suspect at best but it will only wreck your world if you ask for it. 😆

The other nice part about that is that it makes it easier for someone to dive in and perhaps fix future issues.  I say that in the hopes that I never have to look at this file again.  (kidding... sorta).

Anyways, please give this a try whenever you get the chance and let me know how well (or not) it works for you. 👍 

Preview feed is here:
https://www.myget.org/F/wojtpl2/api/v3/index.json
